mumble-voip/mumble
# Stereo stream decoding and playing support

Just executed the idea in #4193. Merging this PR will close #2829, ending a three-year-long pursuit for the stereo audio, realizing the dream of dozens of forks :D.
Thank @Krzmbrzl, @streaps for their help, and @felix91gr for supporting me!

In this commit, I initialize opus decoder in stereo mode. For incoming mono stream, the opus decoder will automatically convert it into a stereo stream so we don't have to make any changes to the protocol, thus this improvement is backward compatible. I carefully tweaked the jitter buffer and the resampler and the fade in/out, and I guess that's all I should take care of.

Actually, opus decoder can also automatically convert a stereo stream into a mono stream, so even if we transmit a stereo stream to an older version of mumble, it would still work.

When mapping the stereo stream to the audio output buffer, I used the `fSpeaker ` direction vector designed for positional audio. It looks like, for CoreAudio, the only two speakers enabled are the front left speaker and front right speaker, and their direction vector does not completely point to left or right, so the mixed audio is not so "stereo". I have to switch on the "Headphone" mode in the settings panel to make that vector completely pointing to left or right.

I didn't change the AudioInput part, partly because it is useless for the normal user who has only one microphone. In fact, the motivation for me is to have mumble decode the stereo stream from my music bot :), so I don't really care about input.

## How to test this PR

To test this PR, you need to have a music bot that supports stereo audio. I have already added the stereo feature to the python library pymumble, and created [a fork of botamusique](https://github.com/TerryGeng/botamusique/tree/stereo) based on this improved version of pymumble. 

Just clone it, checkout the stereo branch `git checkout stereo` and follows the quick start guide to have it installed. Remember to run `git submodule init && git submodule update` to download the improved version of pymumble. Then find some stereo songs and enjoy! And you may also need to turn on "headphone" mode in the settings panel to get the best result.

BTW I just discovered compiled binaries of this PR can be found at [Azure Pipelines](https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=1820&view=artifacts&type=publishedArtifacts). I think people interested having stereo support before next release can check this out.
-----
Closes #2829
Closes #4210
